sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/UploadBufferPool.java

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.blob;
+
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.storage.common.Utility;
+import reactor.core.publisher.Flux;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * This type is to support the implementation of buffered upload only. It is mandatory that the caller
+ * has broken the source into ByteBuffers that are no greater than the size of a chunk and therefore a buffer in the
+ * pool. This is necessary because it upper bounds the number of buffers we need for a given call to write() to 2. If
+ * the size of ByteBuffer passed into write() were unbounded, the pool could stall as it would run out of buffers before
+ * it is able to return a result, and if it is unable to return, no data can be uploaded and therefore no pools
+ * returned.
+ *
+ * It is incumbent upon the caller to return the buffers after an upload is completed. It is also the caller's
+ * responsibility to signal to the pool when the stream is empty and call flush to return any data still sitting in
+ * the pool.
+ *
+ * Broadly, the workflow of this operation is to chunk the source into reasonable sized pieces. On each piece, one
+ * thread will call write on the pool. The pool will grab a buffer from the queue to write to, possibly waiting for one
+ * to be available, and either store the incomplete buffer to be filled on the next write or return the filled buffer to
+ * be sent. Filled buffers can be uploaded in parallel and should return buffers to the pool after the upload completes.
+ * Once the source terminates, it should call flush.
+ */
+final class UploadBufferPool {
+    private final ClientLogger logger = new ClientLogger(UploadBufferPool.class);
+
+    /*
+    Note that a blocking on a syncrhonized object is not the same as blocking on a reactive operation; blocking on this
+    queue will not compromise the async nature of this workflow. Fluxes themselves are internally synchronized to ensure
+    only one call to onNext happens at a time.
+     */
+    private final BlockingQueue<ByteBuffer> buffers;
+
+    private final int maxBuffs;
+
+    private int numBuffs = 0;
```

Comment:
I don't think so because we only have one constructor that takes this as a required parameter. I initialized numBuffs here because it's an internal detail. If you think it's cleaner style, I can initialize both in the constructor or default maxBuffs to 0 even though it'll always get overwritten

Reply:
I think it is fine to leave as is, it's internally only anyways.